src/check.rs

```diff
@@ -60,17 +100,128 @@ pub fn is_available(url: &Url, ctx: &CheckContext) -> Result<(), CheckError> {
+    Key = { link.to_string() };
+    fn fragments_from(
+        link: &Link,
+        fetch_html: impl Fn() -> Result<String, CheckError>
+    ) -> Result<HashSet<String>, CheckError> = {
+        fetch_html().map(|html| parse_fragments(&html))
+    }
+}
+
+fn is_fragment_available(
+    link: &Link,
+    fragment: &str,
+    html: impl Fn() -> Result<String, CheckError>,
+) -> Result<(), CheckError> {
+    let fragments = fragments_from(link, html)?;
+
+    if fragments.contains(fragment) {
+        return Ok(());
+    }
+
+    // Rust documentation uses `#n-m` fragments and JavaScript to highlight
+    // a range of lines in HTML of source code, an element with `id`
+    // attribute of (literal) "#n-m" will not exist, but elements with
+    // `id`s n through m should, this parses the ranged n-m anchor and
+    // checks if elements with `id`s n through m do exist
+    static RUST_LINE_HIGLIGHT_RX: Lazy<Regex> =
+        Lazy::new(|| Regex::new(r#"^(?P<start>[0-9]+)-(?P<end>[0-9]+)$"#).unwrap());
+    match RUST_LINE_HIGLIGHT_RX.captures(fragment) {
+        Some(capture) => match (capture.name("start"), capture.name("end")) {
+            (Some(start_str), Some(end_str)) => {
+                // NOTE: assumes there are less than 2.pow(32) lines in a source file
+                let start = start_str.as_str().parse::<i32>().unwrap();
+                let end = end_str.as_str().parse::<i32>().unwrap();
+                let missing = (start..=end)
+                    .map(|i| i.to_string())
+                    .filter(|i| !fragments.contains(i))
+                    .collect::<Vec<String>>();
+                if !missing.is_empty() {
+                    Err(CheckError::Fragment(
+                        link.clone(),
+                        fragment.to_string(),
+                        Some(missing),
+                    ))
+                } else {
+                    Ok(())
+                }
+            }
+            _ => unreachable!("if the regex matches, it should have capture groups"),
+        },
+        None => Err(CheckError::Fragment(
+            link.clone(),
+            fragment.to_string(),
+            None,
+        )),
+    }
+}
 
 /// Check a URL with the "file" scheme for availability. Returns `false` if it is unavailable.
 fn check_file_url(url: &Url, _ctx: &CheckContext) -> Result<(), CheckError> {
     let path = url.to_file_path().unwrap();
 
-    if path.is_file() || path.join("index.html").is_file() {
-        debug!("Linked file at path {} does exist.", path.display());
-        Ok(())
+    // determine the full path by looking if the path points to a directory,
+    // and if so append `index.html`, this is needed as we'll try to read
+    // the file, so `expanded_path` should point to a file not a directory
+    let expanded_path = if path.is_file() {
+        path.clone()
+    } else if path.is_dir() && path.join("index.html").is_file() {
+        path.join("index.html")
```

Comment:
This `clone()` is not necessary.
```suggestion
    // the file, so `expanded_path` should point to a file not a directory
    let index_html;
    let expanded_path = if path.is_file() {
        &path
    } else if path.is_dir() && path.join("index.html").is_file() {
        index_html = path.join("index.html");
        &index_html
```
and then update `&expanded_path` -> `expanded_path` below.